### PR TITLE
MAGN-7842 Presets that don't contain input nodes should not be allowed

### PR DESCRIPTION
**Pull request by ramramps**
### Purpose

This PR has the changes that  allow preset creation only on input nodes.

Link to YouTrack:
http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-7842
### Declarations

- [x] The code base is in a better state after this PR
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files

### Reviewers

@pboyer

### CODE REVIEW

File: test/DynamoCoreWpfTests/PresetViewModelTest.cs [hunk 1]
```diff
@@ -58,5 +58,37 @@ public void TogglePresetOptionVisibility()
             //Check the Preset option visibility.
             Assert.AreEqual(false, ViewModel.EnablePresetOptions);
         }
+
+        [Test]
+        public void CanCreatePreset()
```
Comment (holyjewsus) [inline test/DynamoCoreWpfTests/PresetViewModelTest.cs]: is this test named correctly?

Comment (ramramps) [inline test/DynamoCoreWpfTests/PresetViewModelTest.cs]: Yes. This is to test the Preset Creation on ViewModel.

Comment (holyjewsus) [inline test/DynamoCoreWpfTests/PresetViewModelTest.cs]: sorry if I'm mistaken here, but where does this test check for preset creation? maybe the diff is incorrect on github, seems to be checking for the GetInputNodes() method

Comment (ramramps) [inline test/DynamoCoreWpfTests/PresetViewModelTest.cs]: Yes. Preset Creation is checked on the View. So, this test checks whether the passed node is an input node. Similiar tests on the model does not have this filter, as it is assumed that only input nodes are passed to create preset.

File: src/DynamoCore/Models/DynamoModel.cs [hunk 2]
```diff
@@ -79,6 +79,13 @@ internal void OnWorkspaceSaved(WorkspaceModel model)
                 WorkspaceSaved(model);
         }
 
+        public event Action RequestPresetsWarningPrompt;
+        public void OnRequestPresetWarningPrompt()
```
Comment (pboyer) [inline src/DynamoCore/Models/DynamoModel.cs]: Can this be `internal`?

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 3]
```diff
@@ -1404,17 +1404,32 @@ private bool CanShowNewFunctionDialogCommand(object parameter)
         /// </summary>
         private void ShowNewPresetStateDialogAndMakePreset(object parameter)
         {
-            //trigger the event to request the display
-            //of the preset name dialogue
-            var args = new PresetsNamePromptEventArgs();
-            this.Model.OnRequestPresetNamePrompt(args);
-            var IDS = DynamoSelection.Instance.Selection.OfType<NodeModel>().Select(x => x.GUID).ToList();
-            if (args.Success)
+            var selectedNodes = GetInputNodes().ToList();
```
Comment (pboyer) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs]: You're calling `ToList()` twice on the same collection.

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 4]
```diff
@@ -1422,6 +1437,12 @@ private bool CanShowNewPresetStateDialog(object parameter)
             return DynamoSelection.Instance.Selection.Count > 0;
         }
 
+        public IEnumerable<NodeModel> GetInputNodes()
+        {
+            return DynamoSelection.Instance.Selection.OfType<NodeModel>()
+                                .Where(x => x.IsInputNode).ToList();
```
Comment (pboyer) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs]: If you're making the type `IEnumerable`, don't return a `List`.

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 5]
```diff
@@ -1404,17 +1404,32 @@ private bool CanShowNewFunctionDialogCommand(object parameter)
...
+
+            //If there are NO input nodes then show the error message
+            if (!selectedNodes.Any())
+            {
+                this.Model.OnRequestPresetWarningPrompt();
+            }
+            else
             {
+                //trigger the event to request the display
+                //of the preset name dialogue
```
Comment (pboyer) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs]: Isn't this obvious?

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 6]
```diff
@@ -1404,17 +1404,32 @@ private bool CanShowNewFunctionDialogCommand(object parameter)
...
+                var args = new PresetsNamePromptEventArgs();
+                this.Model.OnRequestPresetNamePrompt(args);
+
+                //Select only Input nodes for preset
+                var IDS = selectedNodes.Select(x => x.GUID).ToList();
```
Comment (pboyer) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs]: A style thing: It's strange to name a local variable `IDS` in all caps, this doesn't follow .NET style guides. Shouldn't this be `ids`?

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 7]
```diff
@@ -1422,6 +1437,12 @@ private bool CanShowNewPresetStateDialog(object parameter)
             return DynamoSelection.Instance.Selection.Count > 0;
         }
 
+        public IEnumerable<NodeModel> GetInputNodes()
```
Comment (pboyer) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs]: This method is badly named. It doesn't give any info about from which collection the nodes are drawn from. Are we talking about nodes in the current selection? I'd consider inlining it or making it `private` and making it more descriptive.

**Comment by ramramps** [comment]
@pboyer - PTAL now.

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 8]
```diff
@@ -1422,6 +1435,16 @@ private bool CanShowNewPresetStateDialog(object parameter)
             return DynamoSelection.Instance.Selection.Count > 0;
         }
 
+        /// <summary>
+        /// Gets the selected "input" nodes
+        /// </summary>
+        /// <returns></returns>
```
Comment (pboyer) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs]: No need for an empty returns field, but this isn't that big of a deal.

**Pull request merged**